### ISSUE TITLE
Consider splitting the `core` module into `core`, `wallet`, and `net` modules

### ISSUE THREAD

**Issue by msgilligan** [bitcoinj/bitcoinj#1874]
It _might_ make sense to split the `core` module into two modules the new module being called something like `base` and ideally having zero dependencies besides the JRE, but perhaps might end up having 1 or 2 dependencies (e.g. protobuf or guava) if unavoidable.

**NOTE:** The proposal has changed to three modules, and the title has been changed. The above is the original proposal unmodified.

**Comment by msgilligan** [comment]
As @Polve said over in issue #1877, this module would contain:

>  just crypto stuff and parsing classes for base bitcoin objects: no network code or logic to handle the blockchain and wallets.

**Comment by schildbach** [comment]
I actually would call it 'core', and try to find other names for the stuff that we exclude. As a step further out, we could also extract a 'wallet' module.

**Comment by msgilligan** [comment]
Yeah, `core` was my first thought, but it seems like the migration would be easier if we called it something else, so we could start moving code into that module gradually as we refactor. That's why I suggested `base`  (which mirrors `java.base` and `javafx.base`)  Backward compatibility would be provided by `core` depending on `base`.

While refactoring we could create an `org.bitcoinj.base` package (with sub packages like `org.bitcoinj.base.hd`, for example) and leave it in the `core` JAR and automatic JPMS module. Later we could put the `org.bitcoinj.base` package in its own JAR and JPMS module.

So we could start with the `hd` sub package [as we discussed](https://github.com/bitcoinj/bitcoinj/pull/1877#issuecomment-501605765).

What do you think?

**Comment by msgilligan** [comment]
I made an attempt at extracting a wallet module (well just refactoring packages within the existing module to eliminate circular dependencies at the package level by introducing a _temporarily-named_ `WalletIF` class)

See PR #2222

**Comment by msgilligan** [comment]
I changed the title of the issue to match my (our*) latest thinking. 

I think we could break into three modules as follows:

`core`: most of existing "core", defines wallet and network API but not implementation. Looses dependencies on okHttp and (hopefully) protobuf.
`wallet`: current protobuf-based wallet implementation, protobuf dependency
`net`: current networking implementation, okhttp and other networking clients using older JDK APIs

This would allow us to develop experimental implementations of updated wallet and network modules, perhaps one or more of the following:

1. A SQLite-based wallet (perhaps even file compatible with the SQLite-based wallet files in Bitcoin Core)
2. A reimplemented network layer using Java 8 APIs
3. A reimplemented network layer using Java 11 APIs (e.g. `java.net.http`)

They could be experimental initially, but if/when they get stable they can be documented as stable and begin to see actual use.

**bitcoinj** applications would be able to choose which network and wallet implementations (if any) they want to use.

* This has certainly been influenced by many conversation I've had with @schildbach

**Comment by msgilligan** [comment]
Looking at the conversation in PR #1877, it was proposed that blockchain stores be moved out of `core`.  That's probably a good idea, too, but I currently see it as less urgent.

**Comment by msgilligan** [comment]
> three modules as follows:
> 
> `core`: most of existing "core", defines wallet and network API but not implementation. Looses dependencies on okHttp and (hopefully) protobuf.
>`wallet`: current protobuf-based wallet implementation, protobuf dependency
>`net`: current networking implementation, okhttp and other networking clients using older JDK APIs

I think I would add `crypto` as a 4th module.

@schildbach What are your latest thoughts?

**Comment by msgilligan** [comment]
Closing. See https://github.com/bitcoinj/bitcoinj/issues/1874 for discussion of making a more modular **bitcoinj** with multiple crypto implementations.

**Comment by schildbach** [comment]
My gut feeling is that we should probably just start with separating out `wallet`. I mean just move it out; don't introduce a Wallet interface for now.

**Comment by msgilligan** [comment]
> My gut feeling is that we should probably just start with separating out `wallet`. I mean just move it out; don't introduce a Wallet interface for now.

Separating wallet seems like a good place to start. I think we will *need* to create a Wallet interface in order to do it, because we'll need to break circular dependencies between packages/modules.

**Comment by msgilligan** [comment]
So we are making good progress on moving classes into the new `org.bitcoinj.base` package (which in a future release will likely become its own module/JAR.) As of today, the following significant classes have been moved to `org.bitcoinj.base`:

* `Base58`
* `Bech32`
* `Network`/`BitcoinNetwork` (new classes, replacements for the `NetworkParameters` hierarchy for many use cases)
* `Coin`
* `ScriptType`
* `Sha256Hash`

(this is partial list focusing on "significant" classes.)

**Comment by msgilligan** [comment]
We have also been doing some significant refactoring to prepare `Address`/`LegacyAddress`/`SegwitAddress` for movement. Part of this work has been migrating `Address` and subclasses to use `Network/BitcoinNetwork` rather than the `NetworkParameters` hierarchy and this work is nearing completion.

The steps that remain are:

- [ ] Merge PR #2608
- [ ] Merge PR #2609 
- [ ] Merge PR #2610
- [ ] Rebase and merge PR #2501
- [ ] Deprecate usage of `NetworkParameters` in `Address` and subclasses and replace with `Network`/`BitcoinNetwork` (I have some work-in-progress for this that I have not pushed yet)
- [ ] Replace `NetworkParameters.getP2SHHeader()`, `NetworkParameters.getSegwitAddressHrp()` with something we can get directly or indirectly from `Network/BitcoinNetwork`
- [ ] Rebase or replace PR #2604

**Comment by msgilligan** [comment]
I've updated the title and the initial comment to reflect my (our) latest thinking.

PR #2279 removed the OkHttp dependency which was one of the reasons for wanting to create a separate `net` module.

It continues to look like a separate `crypto` module that depends upon `base` and adds Bouncy Castle and includes our `ECKey` class might make sense. We are not committed to doing this yet, but it is under consideration.

`core` would retain everything else except what is factored into a `wallet` module.  (This is the same as previously discussed.)

But, for now, we're focused on creating the `org.bitcoinj.base` package with an eye towards it becoming a separate module in a future release.

**Comment by msgilligan** [comment]
This diagram shows the dependency structure of this proposed architecture: https://github.com/bitcoinj/bitcoinj/blob/master/designdocs/modular-architecture.md

**Comment by rotilho** [comment]
I'm not sure if this is possible, but would it be feasible to remove Guava from the base or core entirely? From what I've gathered, introducing a custom checkArgument and using Objects.requireNonNull might allow you to eliminate the Guava dependency in most places.

**Comment by schildbach** [comment]
> I'm not sure if this is possible, but would it be feasible to remove Guava from the base or core entirely? From what I've gathered, introducing a custom checkArgument and using Objects.requireNonNull might allow you to eliminate the Guava dependency in most places.

I think this is indeed a goal, although not particularly urgent. See #2000.

**Comment by msgilligan** [comment]
> I'm not sure if this is possible, but would it be feasible to remove Guava from the base or core entirely?

I would like to see this happen and have been working (part-time) towards this goal for a few years now. I think removing the Guava dependency in the WIP `base` module is achievable in the near future if we allow ourselves to duplicate (reinvent) some Guava functionality (as you propose.)

I am more gung-ho about doing this than Andreas is, but I think we might be able to convince him if the Guava duplication in `base` is minimal, clean, and can also be used to replace the corresponding Guava usage in the `core` module.

**Comment by msgilligan** [comment]
We have just merged some architecture diagrams in the `designdocs` folder that show the planned refactorings/migrations:

https://github.com/bitcoinj/bitcoinj/blob/master/designdocs/modular-architecture.md

**Comment by msgilligan** [comment]
I closed PR #2222 because it is very stale, though it may be worth referencing for the next attempt at extracting a Wallet interface.

**Comment by msgilligan** [comment]
I created related Issue #3352 which should help us with this migration.

**Comment by msgilligan** [comment]
I created issue #3389 for the migration to `secp256k1-jdk` which is closely related to this issue.

**Comment by msgilligan** [comment]
As of today, PR #3471 is merged to `master`. This means `base` has been migrated to its own module and this new module has _effectively_ zero dependencies.
(I'm not counting `org.slf4j:slf4j-api` or `com.google.code.findbugs:jsr305` as dependencies.)

**Comment by msgilligan** [comment]
I created Issue #3524 for the sub-task of creating a wallet `interface`.  (In that issue I suggest that a wallet interface is not a top priority but still something that should be worked on in the 0.18/0.19 time-frame)

**Comment by msgilligan** [comment]
I added Issue #3912 to look at completely abstracting our usage of Bouncy Castle.
